# Simulation of Nitflex Business Model

Nitflex is a streaming company providing paid content on a subscription-based plan.
The company model generates revenues exclusively through user’s subscription which comprises the following plans:
•	Free Trial (14 days): users subscribe to try the Nitflex services free of charge for a week. The user must provide a credit card during the sign-up process. The user can cancel the free trial at any time during the 14 days; after that, the user is automatically moved to the Basic plan and charged for the first month.
•	Basic Plan: users in the Basic Plan have access to the entire media collection; however, they have a limit of 40 movies a month.
•	Unlimited Plan: users in the Unlimited Plan have access to the entire media collection with no limit.


MM with the transition probabilities

In [8]:
import numpy as np
import random as rm

# Defining the States
states = ['Free Trial', 'Basic Plan', 'Unlimited Plan', 'Lost']
#The four possible states that are made available in the data

# Defining the transitions
transitionName = [['T11', 'T12', 'T13', 'T1D'],
                  ['T21', 'T22', 'T23', 'T2D'],
                  ['T31', 'T32', 'T33', 'T3D'],
                  ['TD1', 'TD2', 'TD3', 'TDD'],
                  ['TL1', 'TL2', 'TL3', 'TLD']]

#All possible transitions, such as Trial to Basic, Basic to Unlimited, Unlimited to Lost, etc

# Probability or Transition Matrix
tMatrix = [[0.0, 30/65, 13.5/65, 21.5/65],
           [0.0, 3/15, 0.0, 12/15],
           [0.0, 9/20, 2.5/20, 8.5/20],
           [0.0, 0.0, 0.0, 1.0]]

#All Trial is never returned to, hence why first column is all 0, rest represents the amount of users that made transitions
#The lost state never transtions to another state which is why it is 1.0
#The data does not state any user going directly from the Basic membership and upgrading to the Unlimited, only upgrading from the trial membership
#To each of the others or downgrading from the Unlimited to the basic, but not basic to unlimited.

The following functions will check the matrix.

In [9]:
def mult_matrices(m1, m2):
    res = [[0 for x in range(len(m1))] for y in range(len(m2))]
    for i in range(len(m1)):
        for j in range(len(m2[0])):
            for k in range(len(m2)):
                res[i][j] = round(res[i][j] + m1[i][k] * m2[k][j], 3)
    return res

def checkMatrix(M):
    # The raw probabilities should sum to 1
    n = len(M)
    psum = 0
    result = False

    for i in range(0,n):
        psum += sum(M[i])

    if psum == n:
        result = True
    return result

In [10]:
# Checking the probabilities look ok: out probabilities should be equal to 1
if checkMatrix(tMatrix):
    print("Probability Test ... passed.")
else:
    print("The transition matrix has some problem")

Probability Test ... passed.


In [11]:
def simulate(periods, starting_state, states, transitionName, tMatrix):
    current_state = starting_state
    transition_index = 0
    projectedMatrix = tMatrix

    transitionList = [states[current_state]]
    period_counter = 1
    # Probability of transitionList
    prob = 1

    while period_counter != periods:

        next_transition = np.random.choice(transitionName[transition_index],replace=True,p=tMatrix[transition_index])
        next_state_index = transitionName[transition_index].index(next_transition)

        if tMatrix[transition_index][next_state_index]>0.0:
            prob = prob * tMatrix[transition_index][next_state_index]

        transition_index = next_state_index
        transitionList.append(states[transition_index])

        projectedMatrix = mult_matrices(projectedMatrix,projectedMatrix)

        period_counter += 1
    return transitionList, states[transition_index], prob, projectedMatrix

## What would be the distribution of probabilities for each Subscription level 10 years from now?


In [12]:
periods = 10
starting_state = 0
transition_list, final_state, probability, projected_matrix = simulate(periods, starting_state, states, transitionName, tMatrix)

print("Possible states: " + str(transition_list))
print("End state after "+ str(periods) + " periods: " + final_state)
print("Probability of the possible sequence of states: " + str(probability))
print()
#Printing projected matrix
for i in range(len(projected_matrix)):
    print(projected_matrix[i])

Possible states: ['Free Trial', 'Lost', 'Lost', 'Lost', 'Lost', 'Lost', 'Lost', 'Lost', 'Lost', 'Lost']
End state after 10 periods: Lost
Probability of the possible sequence of states: 0.33076923076923076

[0.0, 0.0, 0.0, 0.999]
[0.0, 0.0, 0.0, 1.0]
[0.0, 0.0, 0.0, 0.999]
[0.0, 0.0, 0.0, 1.0]


Write your comments in the next cell.

End state after 10 periods will be 100% lost.
Probability of the possible sequence of states: 0.33076923076923076

Based on the data that we have been given, 100% of users will arrive at the lost state in ten years. A large part of the problem potentially occurs in the fact that once a user is lost they never return. This model not allowing users to return if they decide not to renew means that in all likelyhood users will not continue to use the service long term. This service also does not incentivize the basic option since the data does not explicitly express that any of the 100 users in the sample ever decided to upgrade, however it does specificy that users did downgrade. It would be reasonable to assume that some users who liked the service would attempt to upgrade to a more premium version. Not allowing users the ability to gain access to more features unless they do directly from the free trial is a poor way to retain users and may likely be the main reason that the model predicts very confidently that 100% of basic users will be lost, rather then the small possibility that the unlimited option may retain a very small amount.

## What Modification would you introduce for Nitflex to be more Successful?

In [13]:
tMatrix = [[0.0, 32/65, 12.5/65, 20.5/65],
           [0.0, 2/15, 2/15, 11/15],
           [0.0, 9/20, 2.5/20, 8.5/20],
           [0.0, 35/100, 15/100, 50/100]]

#Changes in first row: Allowing their to be an option of going from a basic to an unlimited would likely effect what people do in the trial state.
#It can be reasonably assumed that at least one person would decide differently and go from the trial directly to the basic state, which would effect going from trial
#to unlimited or trial to lost

#Changes in Second Row: Allowing there to be an option to go from the basic plan to the unlimited, We can make similar assumptions that we did with the first row.
#If given the option we can assume that at least one person would decide to choose the to upgrade from basic to unlimited rather then to remain in the basic plan or to
#not renew altogether.


#Changes in Third Row: The changes that we have predicited would likely not effect this row significantly, so I chose to keep it the same

#Changes in Fourth Row: Assumes that Lost state is no longer an absorbant state, meaning users can now return. We can assume that at some point that if users who
#can cancel the service are allowed to renew it again then its possible every user would at some point cancel it and return. I believe that potentially half
#Of all users may end up leaving for good, however if people were to return the larger majority would likely go to the basic state rather then going directly from the
#lost state to the more expensive option. They also cannot return to the trial state since coming from the lost state would mean they were already eligible for the trial.

Run the simulation again (on the same 10 periods) and generate the new simulated transition probabilities.  
  Execute the simulation in the next cell and print the results.

In [14]:
periods = 10
starting_state = 0
transition_list, final_state, probability, projected_matrix = simulate(periods, starting_state, states, transitionName, tMatrix)

print("Possible states: " + str(transition_list))
print("End state after "+ str(periods) + " periods: " + final_state)
print("Probability of the possible sequence of states: " + str(probability))
print()
#Printing projected matrix
for i in range(len(projected_matrix)):
    print(projected_matrix[i])

Possible states: ['Free Trial', 'Basic Plan', 'Unlimited Plan', 'Basic Plan', 'Unlimited Plan', 'Lost', 'Unlimited Plan', 'Basic Plan', 'Unlimited Plan', 'Basic Plan']
End state after 10 periods: Basic Plan
Probability of the possible sequence of states: 6.779076923076922e-06

[0.0, 0.442, 0.211, 0.825]
[0.0, 0.442, 0.211, 0.825]
[0.0, 0.442, 0.211, 0.825]
[0.0, 0.444, 0.211, 0.827]


End state after 10 periods: Basic Plan
Probability of the possible sequence of states: 0.00034150240384615377

This model shows that after ten years Just over 80% of consumers will arrive at the lost state. While This is still a a lot of people as this show between 82 and 83 percent of users will eventually be in the lost state. However, this is still a significant improvment since this is nearly a 20% increase in the amount of active users then previously. These are also conservative observations as it makes the assumption that 50 percent of users will be lost and lost forever despite being able to return to either the basic plan or unlimited plan. It can even be assumed that because the new changes involve being able to upgrade from the basic option to the unlimited that more people may return to the basic state, which can lead to more people in the unlimited state and ultimately lead to more retention. But we can conclude that allowing users to return from the lost state leads to a signifcant improvement in user retention.